File: agent/tools/market_price.go
package tools

import (
	"fmt"
	"net/url"
	"strings"
	"time"

	"github.com/vultisig/vultisig-win/agent/shared"
	"github.com/vultisig/vultisig-win/agent/tools/rpc"
)

const marketPriceBaseURL = "https://api.vultisig.com/coingeicko/api/v3/simple/price"

type GetMarketPriceTool struct{}

func NewGetMarketPriceTool() *GetMarketPriceTool {
	return &GetMarketPriceTool{}
}

func (t *GetMarketPriceTool) Name() string {
	return "get_market_price"
}

func (t *GetMarketPriceTool) Description() string {
	return "Get the latest market price for a cryptocurrency in a fiat currency (default USD). Supports common tickers like BTC/ETH/SOL and CoinGecko IDs."
}

func (t *GetMarketPriceTool) InputSchema() map[string]any {
	return map[string]any{
		"asset": map[string]any{
			"type":        "string",
			"description": "Asset symbol or id (e.g., 'BTC', 'ETH', 'bitcoin', 'usd-coin')",
		},
		"fiat": map[string]any{
			"type":        "string",
			"description": "Fiat quote currency (default: 'usd'), e.g. 'usd', 'eur', 'gbp'",
		},
	}
}

func (t *GetMarketPriceTool) RequiresPassword() bool {
	return false
}

func (t *GetMarketPriceTool) RequiresConfirmation() bool {
	return false
}

func (t *GetMarketPriceTool) Execute(input map[string]any, ctx *ExecutionContext) (any, error) {
	assetRaw, ok := input["asset"]
	if !ok || assetRaw == nil {
		return nil, fmt.Errorf("asset is required")
	}

	asset, ok := assetRaw.(string)
	if !ok {
		return nil, fmt.Errorf("asset must be a string")
	}
	asset = strings.TrimSpace(asset)
	if asset == "" {
		return nil, fmt.Errorf("asset is required")
	}

	fiat := "usd"
	if fiatRaw, exists := input["fiat"]; exists && fiatRaw != nil {
		fiatStr, ok := fiatRaw.(string)
		if !ok {
			return nil, fmt.Errorf("fiat must be a string")
		}
		fiatStr = strings.TrimSpace(strings.ToLower(fiatStr))
		if fiatStr != "" {
			fiat = fiatStr
		}
	}

	priceProviderID := t.resolvePriceProviderID(asset, ctx)

	q := url.Values{}
	q.Set("ids", priceProviderID)
	q.Set("vs_currencies", fiat)

	endpoint := marketPriceBaseURL + "?" + q.Encode()

	var resp map[string]map[string]float64
	err := rpc.GetJSON(endpoint, &resp)
	if err != nil {
		return nil, fmt.Errorf("failed to fetch market price: %w", err)
	}

	coinData, ok := resp[strings.ToLower(priceProviderID)]
	if !ok {
		coinData, ok = resp[priceProviderID]
	}
	if !ok {
		return map[string]any{
			"found":             false,
			"asset":             asset,
			"price_provider_id": priceProviderID,
			"fiat":              fiat,
			"message":           fmt.Sprintf("No market price available for '%s'", asset),
			"source":            "api.vultisig.com/coingeicko/api/v3/simple/price",
		}, nil
	}

	price, ok := coinData[strings.ToLower(fiat)]
	if !ok {
		price, ok = coinData[fiat]
	}
	if !ok {
		return map[string]any{
			"found":             false,
			"asset":             asset,
			"price_provider_id": priceProviderID,
			"fiat":              fiat,
			"message":           fmt.Sprintf("No '%s' quote available for '%s'", fiat, asset),
			"source":            "api.vultisig.com/coingeicko/api/v3/simple/price",
		}, nil
	}

	return map[string]any{
		"found":             true,
		"asset":             asset,
		"price_provider_id": priceProviderID,
		"fiat":              fiat,
		"price":             price,
		"source":            "api.vultisig.com/coingeicko/api/v3/simple/price",
		"as_of":             time.Now().UTC().Format(time.RFC3339),
	}, nil
}

func (t *GetMarketPriceTool) resolvePriceProviderID(asset string, ctx *ExecutionContext) string {
	normalized := strings.TrimSpace(strings.ToLower(asset))

	if resolved := shared.ResolveAsset(normalized); resolved != nil && resolved.PriceProviderID != "" {
		return strings.ToLower(resolved.PriceProviderID)
	}

	if ctx != nil && ctx.Vault != nil {
		for _, coin := range ctx.Vault.Coins {
			if strings.EqualFold(coin.Ticker, normalized) && coin.PriceProviderID != "" {
				return strings.ToLower(coin.PriceProviderID)
			}
		}
	}

	if id, ok := commonTickerToPriceProviderID[strings.ToUpper(normalized)]; ok {
		return id
	}

	return normalized
}

var commonTickerToPriceProviderID = map[string]string{
	"BTC":   "bitcoin",
	"ETH":   "ethereum",
	"SOL":   "solana",
	"BNB":   "binancecoin",
	"AVAX":  "avalanche-2",
	"MATIC": "polygon-ecosystem-token",
	"DOT":   "polkadot",
	"XRP":   "ripple",
	"TON":   "the-open-network",
	"TRX":   "tron",
	"DOGE":  "dogecoin",
	"LTC":   "litecoin",
	"BCH":   "bitcoin-cash",
	"USDC":  "usd-coin",
	"USDT":  "tether",
}

File: core/ui/agent/tools/handlers/marketPrice.ts
import { Chain } from '@core/chain/Chain'
import { chainFeeCoin } from '@core/chain/coin/chainFeeCoin'
import { knownTokensIndex } from '@core/chain/coin/knownTokens/index'
import { getCoinPrices } from '@core/chain/coin/price/getCoinPrices'
import { fiatCurrencies, FiatCurrency } from '@core/config/FiatCurrency'

import { resolveAsset } from '../shared/assetResolution'
import type { ToolHandler } from '../types'

function resolvePriceProviderId(
  asset: string,
  coins: { ticker: string; priceProviderId?: string }[]
): string {
  const normalized = asset.trim().toLowerCase()

  const resolved = resolveAsset(normalized)
  if (resolved?.priceProviderId) return resolved.priceProviderId.toLowerCase()

  for (const chain of Object.values(Chain)) {
    const fc = chainFeeCoin[chain]
    if (fc.ticker.toLowerCase() === normalized && fc.priceProviderId) {
      return fc.priceProviderId.toLowerCase()
    }
  }

  for (const chain of Object.values(Chain)) {
    const tokens = knownTokensIndex[chain]
    if (!tokens) continue
    for (const token of Object.values(tokens)) {
      if (token.ticker.toLowerCase() === normalized && token.priceProviderId) {
        return token.priceProviderId.toLowerCase()
      }
    }
  }

  for (const coin of coins) {
    if (coin.ticker.toLowerCase() === normalized && coin.priceProviderId) {
      return coin.priceProviderId.toLowerCase()
    }
  }

  return normalized
}

function isFiatCurrency(s: string): s is FiatCurrency {
  return (fiatCurrencies as readonly string[]).includes(s)
}

export const handleMarketPrice: ToolHandler = async (input, context) => {
  const asset = String(input.asset ?? '').trim()
  if (!asset) throw new Error('asset is required')

  let fiat = 'usd'
  if (input.fiat) {
    fiat = String(input.fiat).trim().toLowerCase() || 'usd'
  }

  const priceProviderId = resolvePriceProviderId(asset, context.coins)
  const fiatCurrency = isFiatCurrency(fiat) ? fiat : ('usd' as FiatCurrency)

  const prices = await getCoinPrices({
    ids: [priceProviderId],
    fiatCurrency,
  })

  const price = prices[priceProviderId.toLowerCase()] ?? prices[priceProviderId]

  if (price === undefined) {
    return {
      data: {
        found: false,
        asset,
        price_provider_id: priceProviderId,
        fiat,
        message: `No market price available for '${asset}'`,
        source: 'api.vultisig.com/coingeicko/api/v3/simple/price',
      },
    }
  }

  return {
    data: {
      found: true,
      asset,
      price_provider_id: priceProviderId,
      fiat,
      price,
      source: 'api.vultisig.com/coingeicko/api/v3/simple/price',
      as_of: new Date().toISOString(),
    },
  }
}
